Assignment4/bertagh-051730600/task07.py

```diff
@@ -0,0 +1,179 @@
+for s,p,o in g.triples((None, RDFS.subClassOf, ns.LivingThing)):
+  print(s)
+
+"""**TASK 7.2: List all individuals of "Person" with RDFLib and SPARQL (remember the subClasses)**
+
+"""
+
+# SPARQL
+print("SPARQL")
+
+q2 = prepareQuery('''
+   SELECT ?Individual  WHERE {
+    ?Subject RDFS:subClassOf* ns:Person.
+  ?Individual RDF:type ?Subject.
+  }
+  ''',
+  initNs = { "RDFS": RDFS,
+            "RDF": RDF,
+             "ns":ns}
+)
+#Visualize the results
+
+for r in g.query(q2):
+  print(r.Individual)
+
+#RDFLib
+print('RDFLib')
+for s,p,o in g.triples((None, RDFS.subClassOf, ns.Person)):
+   for s1,p1,o1 in g.triples((None,RDF.type, s)):
+     print(s1)
+
+for s2,p2,o2 in g.triples((None, RDF.type, ns.Person)):
+  print(s2)
+
+"""**TASK 7.3: List all individuals of "Person" or "Animal" and all their properties including their class with RDFLib and SPARQL. You do not need to list the individuals of the subclasses of person**
+
+"""
+
+# SPARQL
+print("SPARQL")
+
+q3 = prepareQuery('''
+   SELECT  ?Individual ?property WHERE {
+    ?Individual ?property ?value
+    {
+    ?Subject RDFS:subClassOf ns:Person.
+  ?Individual RDF:type ?Subject.
+  }  UNION{
+    ?Subject RDFS:subClassOf* ns:Animal.
+  ?Individual RDF:type ?Subject.
+  }}
+  ''',
+  initNs = { "RDFS": RDFS,
+            "RDF": RDF,
+             "ns":ns}
+)
+#Visualize the results
+
+for r in g.query(q3):
+  print((r.Individual),(r.property))
+
+#RDFLib
+print('RDFLib')
+for s,p,o in g.triples((None, RDF.type, ns.Person)):
+  for s1,p1,o1 in g.triples((s, None, None)):
+   print(s1 + p1)
+for s,p,o in g.triples((None, RDF.type, ns.Animal)):
+  for s1,p1,o1 in g.triples((s, None, None)):
+   print(s1 + p1)
+
+"""**TASK 7.4:  List the name of the persons who know Rocky**"""
+
+# SPARQL
+print("SPARQL")
+
+q4 = prepareQuery('''
+  SELECT  ?Given WHERE {
+    ?Subject foaf:knows ?RockySmith;
+             vcard:Given ?Given.
+  ?RockySmith vcard:FN ?RockySmithFullName.
+  }
+  ''',
+  initNs = { "foaf": FOAF, "vcard": VCARD, "xsd":XSD}
+)
+
+for r in g.query(q4, initBindings = {'?RockySmithFullName' : Literal('Rocky Smith', datatype=XSD.string)}):
+  print(r.Given)
+
+
+#RDFLib
+print('RDFLib')
+for s,p,o in g.triples((None, FOAF.knows, ns.RockySmith)):
+  for s1,p1,o1 in g.triples((s, VCARD.Given , None)):
+    print(o1)
+
+"""**Task 7.5: List the entities who know at least two other entities in the graph**"""
+
+# SPARQL
+print("SPARQL")
+
+q5 = prepareQuery('''
+SELECT ?Entity WHERE {
+  ?Entity foaf:knows ?Entity1.
+  OPTIONAL { ?Entity foaf:knows ?Entity2. FILTER (?Entity1 != ?Entity2) }
+}
+GROUP BY ?Entity
+HAVING (COUNT(?Entity1) >= 2)
+''',
+  initNs = { "foaf": FOAF}
+)
+
+# Visualize the results
+
+for r in g.query(q5):
+  print(r.Entity)
+
+
+#RDFLib
+print('RDFLib')
+
+entities = set()
+
+for s,p,o in g.triples((None, FOAF.knows, None)):
+  for s1,p1,o1 in g.triples((s, FOAF.knows, None)):
+    if o1 != o and s1 not in entities:
+      print(s)
+      entities.add(s1)
```

Comment:
This lists all entities who know someone, but it does not do what the exercise asks for